README.md

```diff
@@ -1,4 +1,219 @@
+
+This allow us to only have to deal with `transmute` at ffi boundaries. Effectively, it allow us to not
+have to rely on the highly `unsafe` `RawPointer` on array implementations, as well as `as_typed` function that transmutes buffers.
+
+[Here](src/buffer/imutable.rs) you can find the concrete implementation proposed in this repo.
+
+### 5. Remove `ArrayData` and place its contents directly on the corresponding arrays.
+
+For example, for primitive types, such as `Float64` and `Date32`, declare a `PrimitiveArrayData<T>` as follows:
+
+```rust
+#[derive(Debug, Clone)]
+pub struct PrimitiveArray<T: NativeType> {
+    data_type: DataType,
+    values: Buffer<T>,
+    validity: Option<Bitmap>,
+    offset: usize,
+}
+```
+
+Note how `T` denotes the _physical_ representation, while `data_type` corresponds to the _logical_ representation. This is so that `Timestamp` with timezones becomes a first class citizen (it currently isn't).
+
+### 6. Child data is stored as `Arc<dyn Array>`
+
+For example, the struct holding a `ListArray` is [defined](src/array/list.rs) as
+
+```rust
+#[derive(Debug, Clone)]
+pub struct ListArray<O: Offset> {
+    data_type: DataType,
+    offsets: Buffer<O>,
+    values: Arc<dyn Array>,
+    validity: Option<Bitmap>,
+    offset: usize,
+}
+```
+
+This greatly simplifies creating nested structures, as there is no longer any `ArrayData`.
+
+Accessing individual (nested) values of this array, e.g. for iterations, work as before:
+
+```rust
+impl<O: Offset> ListArray<O> {
+    pub fn value(&self, i: usize) -> Box<dyn Array> {
+        let offsets = self.offsets.as_slice();
+        let offset = offsets[i];
+        let offset_1 = offsets[i + 1];
+        let length = (offset_1 - offset).to_usize().unwrap();
+
+        self.values.slice(offset.to_usize().unwrap(), length)
+    }
+}
+```
+
+Note the usage of `Array::slice`, an abstract method that each specific implementation must know how to perform. This method has been problematic in the past because its implementation is type-specific, but
+the current implementation is type-agnotisc (i.e. a bug).
+
+In the case of a list array:
+
+```rust
+impl<O: Offset> ListArray<O> {
+    pub fn slice(&self, offset: usize, length: usize) -> Self {
+        let validity = self.validity.as_ref().map(|x| x.slice(offset, length));
+        Self {
+            data_type: self.data_type.clone(),
+            offsets: self.offsets.slice(offset, length),
+            values: self.values.clone(),
+            validity,
+            offset,
+        }
+    }
+}
+
+impl<O: Offset> Array for ListArray<O> {
+    fn slice(&self, offset: usize, length: usize) -> Box<dyn Array> {
+        Box::new(self.slice(offset, length))
+    }
+}
+```
+
+Note how the `offsets` were sliced, but the `values` not. In the current master, both get sliced, which
+is semantically incorrect.
+
+Also note that the choice of `Arc` over `Box` is solely for the purposes of enabling a cheap `Clone`.
```

Comment:
I don't understand this comment -- the example shows the use of a `Box` but the comment suggest that `Arc` is preferred to allow cheap clones

Reply:
You are right: the example is wrong: it is supposed to show an `Arc`.